### PR TITLE
Adding datasource and dataentitties methods to ts sdk

### PR DESCRIPTION
**Pull request by pariosur**
Adding:

- Datasources - add, get, getAll, update, delete
- DataEntity - add ,get, getAll, delete
- addIndex

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -49,6 +49,7 @@
     "typescript": "^4.9.5"
   },
   "dependencies": {
+    "@getmetal/metal-sdk": "^7.1.1",
```
Comment (Czechh) [inline package.json:52]: This makes this package a dependency of itself. I don't think we want that

File: src/metal.ts [hunk 2]
```diff
@@ -299,6 +299,152 @@ export class Metal implements Client {
...
+      },
+    })
+
+    return data
+  }
+
+  async getAllDataSources(limit?: number, offset?: number): Promise<object> {
+    const params = new URLSearchParams()
+    if (limit) params.append('limit', limit.toString())
+    if (offset) params.append('offset', offset.toString())
```
Comment (Czechh) [inline src/metal.ts:338]: There isn't an offset, we have `page` instead.

File: src/metal.ts [hunk 3]
```diff
@@ -299,6 +299,152 @@ export class Metal implements Client {
...
+        'x-metal-api-key': this.apiKey,
+        'x-metal-client-id': this.clientId,
+      },
+    })
+
+    return data
+  }
+
+  async getAllDataSources(limit?: number, offset?: number): Promise<object> {
+    const params = new URLSearchParams()
```
Comment (Czechh) [inline src/metal.ts:336]: @softboyjimbo do you have thoughts on how to add querystrings with fetch?

File: src/metal.ts [hunk 4]
```diff
@@ -299,6 +299,152 @@ export class Metal implements Client {
...
+    return data
+  }
+
+  async getDataEntity(id: string): Promise<object> {
+    if (!id) {
+      throw new Error('dataentity_id required')
+    }
+
+    const url = `${API_URL}/v1/data-entities/${id}`
+    const data = await request(url, {
```
Comment (Czechh) [inline src/metal.ts:397]: We should specify the method even if `GET` is the default

File: src/metal.ts [hunk 5]
```diff
@@ -299,6 +299,152 @@ export class Metal implements Client {
...
+        'x-metal-client-id': this.clientId,
+      },
+    })
+
+    return data
+  }
+
+  async deleteDataEntity(id: string): Promise<void> {
+    if (!id) {
+      throw new Error('dataentity_id required')
```
Comment (Czechh) [inline src/metal.ts:410]: I think `id` would be better here in the error message

File: src/metal.ts [hunk 6]
```diff
@@ -299,6 +299,152 @@ export class Metal implements Client {
...
+  }
+
+  async getAllDataEntities(datasourceId: string, limit?: number, offset?: number): Promise<object> {
+    if (!datasourceId) {
+      throw new Error('datasource ID required')
+    }
+
+    const params = new URLSearchParams()
+    if (limit) params.append('limit', limit.toString())
+    if (offset) params.append('offset', offset.toString())
```
Comment (Czechh) [inline src/metal.ts:435]: No offset, `page` instead https://docs.getmetal.io/api-reference/introduction#normal-pagination

File: src/metal.ts [hunk 7]
```diff
@@ -299,6 +299,152 @@ export class Metal implements Client {
...
+
+    const url = `${API_URL}/v1/datasources/${datasourceId}/data-entities?${params.toString()}`
+    const data = await request(url, {
```
Comment (Czechh) [inline src/metal.ts:438]: We should add the GET method even if it's defaulted

**Review by Czechh — COMMENTED** [review]
Looks great so far! Just some notes:

- We use `page`
- We should specify the method `GET` even if it's defaulted
- Remove the dependency and increase the version number

File: package.json (hunk 1 above)
Comment (pariosur) [reply package.json:52]: right! removing. and from 7.1.1 to 7.1.2?

File: src/metal.ts (hunk 2 above)
Comment (pariosur) [reply src/metal.ts:338]: noted!

File: src/metal.ts (hunk 6 above)
Comment (pariosur) [reply src/metal.ts:435]: noted! will change in python sdk too

File: src/metal.ts (hunk 7 above)
Comment (pariosur) [reply src/metal.ts:438]: nice

File: package.json (hunk 1 above)
Comment (Czechh) [reply package.json]: yes

Comment (Czechh) [reply package.json]: More info about semantic versioning here @pariosur https://www.geeksforgeeks.org/introduction-semantic-versioning/

Comment (pariosur) [reply package.json]: ah nice! so then it would be 7.2.1, no?

**Comment by Czechh** [comment]
Let me know when this is ready for review

**Comment by pariosur** [comment]
> Let me know when this is ready for review

it's ready!

File: package-lock.json [hunk 8]
```diff
@@ -6954,6 +6966,14 @@
         "strip-json-comments": "^3.1.1"
       }
     },
+    "@getmetal/metal-sdk": {
```
Comment (Czechh) [inline package-lock.json:6969]: Looks like this dependency snuck in here. Make sure you remove by doing `npm uninstall @getmetal/metal-sdk`

File: package-lock.json [hunk 9]
```diff
@@ -1,14 +1,15 @@
 {
   "name": "@getmetal/metal-sdk",
-  "version": "7.0.1",
+  "version": "7.1.1",
```
Comment (Czechh) [inline package-lock.json:3]: this should match the package version

File: package-lock.json [hunk 10]
```diff
@@ -1,14 +1,15 @@
...
   "lockfileVersion": 2,
   "requires": true,
   "packages": {
     "": {
       "name": "@getmetal/metal-sdk",
-      "version": "7.0.1",
+      "version": "7.1.1",
```
Comment (Czechh) [inline package-lock.json:9]: this should match the package version

File: package-lock.json [hunk 11]
```diff
@@ -1,14 +1,15 @@
...
       "license": "Apache 2.0",
       "dependencies": {
+        "@getmetal/metal-sdk": "^7.1.1",
```
Comment (Czechh) [inline package-lock.json:12]: this should not be here

File: src/metal.ts [hunk 12]
```diff
@@ -299,6 +299,156 @@ export class Metal implements Client {
 
     return await this.uploadFileToUrl({ url: resource.url, file: fileData, fileType, fileSize })
   }
+
+  async createDataSource(payload: Record<string, any>): Promise<object> {
```
Comment (Czechh) [inline src/metal.ts:303]: The type here `Record<string, any>` is to ambiguous. We need to create a payload type specifically for DataSources so when people are using  the library their code editors can help them understand the schema. If you check other methods there're types added for the payloads.

File: src/metal.ts (hunk 3 above)
Comment (Czechh) [reply src/metal.ts:337]: Still need James' thoughts here

File: src/metal.ts [hunk 13]
```diff
@@ -299,6 +299,156 @@ export class Metal implements Client {
...
+        'x-metal-client-id': this.clientId,
+      },
+    })
+
+    if (!response.ok) {
+      throw new Error(`Error deleting data source: ${response.statusText}`)
+    }
+  }
+
+  async updateDataSource(id: string, payload: Record<string, any>): Promise<object> {
```
Comment (Czechh) [inline src/metal.ts:374]: Same as the Create, we need a defined payload type to help users as `Record<string, any>` is not enough

**Review by Czechh — COMMENTED** [review]
Good progress!

**Comment by Czechh** [comment]
We're missing tests as well for these

**Comment by pariosur** [comment]
@Czechh @softboyjimbo updates ready

> We're missing tests as well for these

what to do here?

File: src/metal.ts (hunk 3 above)
Comment (softboyjimbo) [reply src/metal.ts:339]: this looks good to me!

**Comment by softboyjimbo** [comment]
@pariosur you can reference the tests here - https://github.com/getmetal/metal-ts/tree/main/tests

We'll want to follow these patterns to add tests for the new methods that you created. Typically for code changes like this, you'll need to also add tests to cover them

File: src/metal.ts [hunk 14]
```diff
@@ -321,6 +323,156 @@ export class Metal implements Client {
 
     return await this.uploadFileToUrl({ url: resource.url, file: fileData, fileType, fileSize })
   }
+
+  async createDataSource(payload: CreateDataSourcePayload): Promise<object> {
```
Comment (Czechh) [inline src/metal.ts:327]: Nit: this should be `createDatasource`

File: src/metal.ts [hunk 15]
```diff
@@ -321,6 +323,156 @@ export class Metal implements Client {
...
+        'Content-Type': 'application/json',
+        'x-metal-api-key': this.apiKey,
+        'x-metal-client-id': this.clientId,
+      },
+    })
+
+    return data
+  }
+
+  async getDataSource(id: string): Promise<object> {
```
Comment (Czechh) [inline src/metal.ts:342]: `getDatasource`

File: src/metal.ts [hunk 16]
```diff
@@ -321,6 +323,156 @@ export class Metal implements Client {
...
+        'Content-Type': 'application/json',
+        'x-metal-api-key': this.apiKey,
+        'x-metal-client-id': this.clientId,
+      },
+    })
+
+    return data
+  }
+
+  async getAllDataSources(limit?: number, page?: number): Promise<object> {
```
Comment (Czechh) [inline src/metal.ts:360]: `getAllDatasources`

File: src/metal.ts [hunk 17]
```diff
@@ -321,6 +323,156 @@ export class Metal implements Client {
...
+        'Content-Type': 'application/json',
+        'x-metal-api-key': this.apiKey,
+        'x-metal-client-id': this.clientId,
+      },
+    })
+
+    return data
+  }
+
+  async deleteDataSource(id: string): Promise<void> {
```
Comment (Czechh) [inline src/metal.ts:378]: Ibid

File: src/metal.ts [hunk 18]
```diff
@@ -321,6 +323,156 @@ export class Metal implements Client {
...
+        'x-metal-client-id': this.clientId,
+      },
+    })
+
+    if (!response.ok) {
+      throw new Error(`Error deleting data source: ${response.statusText}`)
+    }
+  }
+
+  async updateDataSource(id: string, payload: UpdateDataSourcePayload): Promise<object> {
```
Comment (Czechh) [inline src/metal.ts:398]: Ibid

File: src/metal.ts [hunk 19]
```diff
@@ -321,6 +323,156 @@ export class Metal implements Client {
...
+        'Content-Type': 'application/json',
+        'x-metal-api-key': this.apiKey,
+        'x-metal-client-id': this.clientId,
+      },
+    })
+
+    return data
+  }
+
+  async getDataEntity(id: string): Promise<object> {
```
Comment (Czechh) [inline src/metal.ts:417]: This one is good! as it's two words

File: src/types.ts [hunk 20]
```diff
@@ -136,3 +136,23 @@ export interface FilePayload {
...
   url: string
 }
+
+export interface MetadataField {
+  name: string
+  type: string | number
+  description: string
+}
+
+export interface CreateDataSourcePayload {
```
Comment (Czechh) [inline src/types.ts:146]: This should use Datasource

File: src/types.ts [hunk 21]
```diff
@@ -136,3 +136,23 @@ export interface FilePayload {
...
+  name: string
+  sourcetype: 'File' | 'Text'
+  autoExtract: boolean
+  metadataFields?: MetadataField[]
+}
+
+export interface UpdateDataSourcePayload {
```
Comment (Czechh) [inline src/types.ts:153]: Ibid

**Review by Czechh — COMMENTED** [review]
Couple of comments

**Comment by pariosur** [comment]
> We need the tests to get this one out, but it's looking great so far!

@Czechh working on them! i just pushed to integrate your recent pr

File: src/metal.ts (hunk 14 above)
Comment (pariosur) [reply src/metal.ts]: cool! this is fixed

**Comment by pariosur** [comment]
@Czechh @softboyjimbo this is ready for review!

**Review by Czechh — APPROVED** [review]

**Review by softboyjimbo — APPROVED** [review]

**Pull request merged**